agent/engine/docker_task_engine.go

```diff
@@ -409,6 +407,39 @@ func (engine *DockerTaskEngine) monitorQueuedTasks(ctx context.Context) {
 	}
 }
 
+func (engine *DockerTaskEngine) tryDequeueWaitingTasks(task *managedTask) bool {
+	// Isolate monitorQueuedTasks processing from changes of desired status updates to prevent
+	// unexpected updates to host resource manager when tasks are being processed by monitorQueuedTasks
+	// For example when ACS StopTask event updates arrives and simultaneously monitorQueuedTasks
+	// could be processing
+	engine.monitorQueuedTasksLock.Lock()
+	defer engine.monitorQueuedTasksLock.Unlock()
```

Comment:
Should this be read lock? I suppose it does not matter as there will be only one instance of `tryDequeueWaitingTasks` but still theoretically I think this should be a `RLock()` and `RUnlock()`

Reply:
Besides, would critical region only be after line 417 (or line 422) since I don't think a task's desired status can change from `stopped` to earlier states? Do we see a potential that things after line 422 takes too much time and holding the lock too long?

Reply:
Once topTask's `taskDesiredStatus` has been read by `tryDequeueWaitingTasks`, it should not be updated anywhere else while this step is in progress. So imo it, the read call itself `task.GetDesiredStatus()` should be in the critical section, hence I declared it within the lock boundary.

The rest of the function will only execute based on the `taskDesiredStatus` read here, so I think it should stay in the critical section

Also, improbable but `failWaitingTask` can change desired status within this function - when the task definition itself has problems. So it I think it might be better to use `Lock` instead of `Rlock` here.

`too much time and holding the lock too long` - I think the method calls are fairly small 
- ToHostResources - this is a method call to map from task struct to host resources
- `dequeueTask` - dequeue from `waitingTasksQueue`
- `hostResourceManager.consume` - updating host resource manager resources - loops over resources and acquires resources if possible
- `task.consumedHostResourceEvent <- struct{}{}` - this is just a write on a buffered channel and returns immediately

Reply:
Thanks! Make sense to me.

Reply:
Oh, one thing is that `consume` itself also seems to acquire a different lock. Not sure if this is a valid case, but I would be cautious about the case where there are both 
- incoming new task placement, which enters the queue and acquire `monitorQueuedTasksLock` and `hostResourceManagerRWLock`
- tasks from restarting the agent, which directly acquires `hostResourceManagerRWLock` without entering the queue

Do we have protection mechanism for the second?

Reply:
Yes, so tasks on restarts pre-acquire resources in [reconcileHostResources](https://github.com/prateekchaudhry/amazon-ecs-agent/blob/taskAccountingAcsStopTask/agent/engine/docker_task_engine.go#L638) before the queue goroutine `monitorQueuedTasks` starts. So this should be taken care of.

Reply:
Great!